@here/harp-mapview/lib/techniques/ExtrudedPolygonTechniqueHandler.ts

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (C) 2020 HERE Europe B.V.
+ * Licensed under Apache 2.0, see full license in LICENSE
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as THREE from "three";
+
+import {
+    BufferAttribute,
+    Expr,
+    ExtrudedPolygonTechnique,
+    Geometry,
+    getPropertyValue,
+    Group
+} from "@here/harp-datasource-protocol";
+import {
+    EdgeMaterial,
+    EdgeMaterialParameters,
+    MapMeshBasicMaterial,
+    MapMeshStandardMaterial
+} from "@here/harp-materials";
+import { assert } from "@here/harp-utils";
+import { AnimatedExtrusionTileHandler } from "../AnimatedExtrusionHandler";
+import {
+    applyBaseColorToMaterial,
+    applySecondaryColorToMaterial,
+    createMaterial,
+    getBufferAttribute
+} from "../DecodedTileHelpers";
+import {
+    createDepthPrePassMesh,
+    isRenderDepthPrePassEnabled,
+    setDepthPrePassStencil
+} from "../DepthPrePass";
+import { Tile } from "../Tile";
+import { techniqueHandlers, TechniqueUpdateContext, TileObjectEntry } from "./TechniqueHandler";
+import { GenericTechniqueHandler, TechniqueHandlerCommon } from "./TechniqueHandlerCommon";
+
+export class ExtrudedPolygonTechniqueHandler extends GenericTechniqueHandler<
+    ExtrudedPolygonTechnique
+> {
+    mainMaterial: MapMeshBasicMaterial | MapMeshStandardMaterial;
+
+    edgeMaterial?: EdgeMaterial;
+
+    mainObjects: TileObjectEntry[] = [];
+    edgeObjects: TileObjectEntry[] = [];
+
+    extrusionAnimationEnabled: boolean = false;
+    extrusionAnimationDuration?: number;
+
+    constructor(technique: ExtrudedPolygonTechnique, tile: Tile, context: TechniqueUpdateContext) {
+        // this.technique = fillDefaults(techniqe)
+        super(technique);
+        this.mainMaterial = createMaterial({ technique, env: context.env }) as
+            | MapMeshBasicMaterial
+            | MapMeshStandardMaterial;
+        assert(
+            this.mainMaterial instanceof MapMeshBasicMaterial ||
+                this.mainMaterial instanceof MapMeshStandardMaterial
+        );
+
+        if (Expr.isExpr(technique.color) || Expr.isExpr(technique.opacity)) {
```

Comment:
Just a generic question about performance since I'm not a JS perf expert:
What is the overhead (memory and speed) of all these callbacks?
We basically add a callback for every dynamic property in every technique.
The alternative would be to have one callback per technique and iterate through all properties.
That would lead to less callbacks but we would pay for the `isExpr` check every time.

So as I said I'm not an expert in JS perf so I cannot judge what would be the better aproach.

Reply:
TLDR: I expected this question and i will try to measure.

Yes, alternative would be one big "update" function that checks what is dynamic, what not.

We have 3 approaches here
1) super update function with "ifs"
2) dedicated callbacks for each property only if needed
3) completly generic updaters like in #1287

I like 2 only because engine will only call these updaters that are needed, albeit taking cost of most probably** dynamic dispatch.

"1" has advantage of static dispatch, but more work and boilerplate.

"3" is dynamic dispatch, correctness but no boilerplate.

Still i think dynamic dispatch can't be so expensive that i choose "2" or better "3" unless measured.

And of course, trying to measure this is really hard because this code (updates of materials) take up to 2-5% of render, so within error range it's really hard to prepare test that will will both isolate this particular functionality and behave like in full application.

Reply:
I agree this is really hard to measure.
Not sure how the JavaScript runtime is actually doing closures(i.e. to know what it should capture it needs to analyse the code when the callback is created). Compared to C++ Lambda where you can exactly specify what and how things should be captured JS closures are a black box to me that support you in shooting yourself into the foot.

Nevertheless I think 2 is a good compromise in code readability and performance.

Reply:
Sorry I just looked at the code again and realised that you are not using closures but bind. So we know what get's captured and everything  is fine.

Reply:
> Not sure how the JavaScript runtime is actually doing closures(i.e. to know what it should capture it needs to analyse the code when the callback is created).

From what i see in V8/Chrome, in debugger at least, it looks like it captures sum of all captured variables in all closures from that scope:

![image](https://user-images.githubusercontent.com/1507542/75031090-c7fff980-54a5-11ea-88b9-20b56493f853.png)

Test:
 breakpoint in onBeforeRenderCallBack for solid lines

Look, that closure captured only `mapView`, `dataSource` and `Tile`. Looking at closures created in this scope, confirms suspicion.

On the other hand i'm not sure if debugger reflects real optimizations in real code emitted by JIT.